### ISSUE TITLE
Allow interpolating based on only the fractional part of the zoom.

### ISSUE THREAD

**Issue by smcallis** [maplibre/maplibre-gl-js#2305]
## User Story

As a user I can interpolate based on the fractional zoom level to preserve the structure in my heatmap on a per-tile basis.

## Rationale

The back story here is that I'm generating a heatmap layer as a fall back when I have too much geometry to send back to the client as a vector tile.  What I do is divide up the tile into 128x128 discrete points and round the bounding box center for each piece of geometry to one of those points.  Then I can return them as a POINT layer and generate a heatmap off of it which provides a reasonably good approximation of the actual geometry underneath.

The problem I'm having is that it's nigh impossible to balance the heatmap radius/intensity etc at a global level doing this, there's no constant value that will work well for everything and interpolating by the global zoom value won't do the right thing.  Effectively what you want do is define some value that's good when you're "right at" the zoom level of the tile, say, 7, and a value that's good when you're almost zoomed past it, say 7.9, and interpolate by that fractional value, then start over at the next integer zoom level (so imagine a saw-tooth pattern).

Right now I don't think this is possible because you can only interpolate on the overall zoom value, not e.g. ["-", "zoom", ["floor", "zoom"]], which is what I want.

**Comment by IvanSanchez** [comment]
If you're dividing the tile into a 128x128 grid... wouldn't you be better off using a 128x128 **raster** tile instead?

**Comment by smcallis** [comment]
No because then I'd have to generate the heatmap myself, and it would be a
lot lower resolution.

**Comment by smcallis** [comment]
Also I'm not sure there's a good way to transition from a vector tile to a raster tile is there?  I'd have add a separate raster endpoint and figure out how to do that blending.  This way I just always return a vector tile and change the layer name depending on if it's a heatmap tile or real geometry.  It works well, just can't style it quite the way I want.

**Comment by HarelM** [comment]
Can you add some illustrations to better explain what you need/expect/want?
I'm having a hard time truly understanding this feature request.

**Comment by smcallis** [comment]
Sure, here's a concrete example, this shows a tile right after it's been loaded, so I'm right on zoom level 7:

![image](https://user-images.githubusercontent.com/804394/227971095-ba145920-716e-4704-a098-88f56c75f770.png)

As I zoom in closer, you can see the image change.  This is right before the level 8 tiles are loaded, so I'm at zoom level like 7.9:

![image](https://user-images.githubusercontent.com/804394/227971311-6719c461-7993-4f73-9afa-d0d252ba6f03.png)

It changes because the points under the heatmap are a fixed distance apart and a fixed radius.

_right now_ I can only modify the radius based on the global zoom level (so e.g. interpolate between zoom 3 and 9 or something like that).  But at each zoom level I load a new set of points that are again equally spaced, so to keep everything looking constant as I zoom in, I need to be able to interpolate on the fractional part of the zoom only.

In this case I'd have some start radius at zoom level N.0 where the tile has just been loaded and takes up 512x512 pixels, and then as you zoom in to level N.9 and it gets closer to 1024x1024, I'd expand the radius, then reset as the new tiles load.

**Comment by HarelM** [comment]
Ah, I see.
I think the original thought behind the heat map what not to receive different data between tile but to get everything up front.
IDK.
I'm not sure what would be the best solution here...

**Comment by smcallis** [comment]
Yeah it makes a lot more sense to configure it globally in the context of the Earthquake data demo, where it's loaded as a single GeoJson.  It works well for what I'm doing here which is summarizing the underlying geometry when it gets too big, it's nice that the heatmap actually reflects the geometry underneath: 

![image](https://user-images.githubusercontent.com/804394/228269937-de50155c-afa9-4991-bd91-cf3efc5f1dac.png)

And as you zoom in the red heatmap blobs kind of shrink-fit themselves to the data and then the real data pops in, it's satisfying, it'd just be nice to have a little more control over the per-tile look of the data.
